src/Sylius/Component/Core/StateResolver/OrderPaymentStateResolver.php

```diff
@@ -44,35 +44,66 @@ public function __construct(FactoryInterface $stateMachineFactory)
     public function resolve(OrderInterface $order)
     {
         $stateMachine = $this->stateMachineFactory->get($order, OrderPaymentTransitions::GRAPH);
-        $completedPaymentTotal = 0;
+        $targetTransition = $this->getTargetTransition($order);
 
-        $payments = $order->getPayments()->filter(function (PaymentInterface $payment) {
-            return PaymentInterface::STATE_COMPLETED === $payment->getState();
-        });
+        if (null !== $targetTransition) {
+            $this->applyTransition($stateMachine, $targetTransition);
+        }
+    }
+
+    /**
+     * @param StateMachineInterface $stateMachine
+     * @param string $transition
+     */
+    private function applyTransition(StateMachineInterface $stateMachine, $transition)
+    {
+        if ($stateMachine->can($transition)) {
+            $stateMachine->apply($transition);
+        }
+    }
 
-        foreach ($payments as $payment) {
+    /**
+     * @param OrderInterface $order
+     *
+     * @return string|null
+     */
+    private function getTargetTransition(OrderInterface $order)
+    {
+        $paymentsCount = $order->getPayments()->count();
+        $refundedPaymentsCount = $this->getPaymentsWithState($order, PaymentInterface::STATE_REFUNDED)->count();
+
+        if ($refundedPaymentsCount === $paymentsCount) {
+            return OrderPaymentTransitions::TRANSITION_REFUND;
```

Comment:
This condition is not true. Among paid payments we can also have `failed` payments and then this condition will never be `true`. We should have almost the same check as with `paid`. This time we should check if the refunded amount is equal to the amount paid.